### DISCUSSION TITLE
Problem with coupled variable in PenaltyInterfaceDiffusion

### DISCUSSION THREAD

**Discussion by Chaoyue-Liu** [idaholab/moose#19551] [Q&A General] [answered]
Hi,
I am trying to modify PenaltyInterfaceDiffusion to implement Butler-Volmer function on an interface. The residual is following:

eta = _u[_qp] - _neighbor_value[_qp];
residual = _F*_k* sqrt(_csmax-_cs[_qp]) * sqrt(_cs[_qp])* sqrt(_cl[_qp])* (exp(C * eta)-exp(-C*eta));

F, k, C are constant. _cs and _cl are two coupled value exist in the subdomain of _u and _neighbor_value respectively. _cs and _cl are calculated by simple diffusion. Somehow when feeding them to the PenaltyInterfaceDiffusion kernel as coupled value, the result of cs and cl is changed. I want to ask how to let PenaltyInterface kernel only consume the cs and cl value without affecting the calculation of them. 

Another question is, how does PenaltyInterfaceDiffusion kernel manage to enforce the diffusion flux equality on the interface without explicitly write it in the residual, like in the InterfaceDiffusion kernel? And where should I modify if I want a different flux formula for PenaltyInterfaceDiffusion?

Thank you!

Chao

**Comment by jiangwen84** [comment, marked as answer]
1. If your cs and cl couple with u, any changes to u will have an effect on cs and cl. 
2. `PenaltyInterfaceDiffusion` does not enforce the flux equality. It simply applies the same flux to either sides of the interface. The residual term is actually the flux you apply.

**Reply by Chaoyue-Liu** [reply]
So how do I decouple cs and cl with u? I just want to feed the value to the expression of u. For now what I do is in the header file declare cs:
const VariableValue _cl;
In the C file:
params.addRequiredCoupledVar("cl");
_cl(coupledValue("cl");
How do I change it?

**Reply by jiangwen84** [reply]
Are they coupled in your PDEs? It is better to list your equation systems here.

**Reply by Chaoyue-Liu** [reply]
The equation on the interface is:
![image](https://user-images.githubusercontent.com/95432912/145259231-402ca7d3-f484-48fb-a007-872106046fb8.png)
![image](https://user-images.githubusercontent.com/95432912/145259395-f9060a05-a048-4e72-b3e2-83de355d8d7f.png)are the variables, others are constants, and i is the flux apply on both sides of the interface.
I gave the cl and cs a good scaling factor (they are far too large comparing to phis and phil), then I have the good result. It's solved.

But I still have another question. Now the PenaltyInterfaceDiffusion assumes that the flux on each side of the interface is :
![image](https://user-images.githubusercontent.com/95432912/145260232-0951185a-76f4-4319-9497-ec2a04985667.png)
I would like to change it to:
![image](https://user-images.githubusercontent.com/95432912/145260473-288799e3-1358-4cd3-8e43-ce9a30219008.png)
where K is a constant. How can I change that?

**Reply by jiangwen84** [reply]
As I suggested before, you should write your own `interfaceKernel` instead of modifying `PenaltyInterfaceDiffusion`.

In your new class, you can apply the residual (which is the flux you gave) to either side of the interface. BTW, you provided the flux in both the first and last equation, which one do you want to use?

**Reply by Chaoyue-Liu** [reply]
Yes I am writing my own interfaceKernel, I'm just using PenaltyInterfaceDiffusion code as a template.

Maybe I'm didn't express myself clear. The PenaltyInterfaceDiffusion kernel deal with following equation set:
![image](https://user-images.githubusercontent.com/95432912/145850036-42699721-b498-498e-9e7b-1a97646fd780.png)
What I want to do is change P(u-u_neighbor) to i, which is:
![image](https://user-images.githubusercontent.com/95432912/145849259-8df111f0-a032-4662-905e-65ddd49a0fcb.png)
where ![image](https://user-images.githubusercontent.com/95432912/145849396-ebd68ecd-97fd-47b1-bf42-533eadef5c25.png) are the variables, others are constants. Also, to change the flux expression ![image](https://user-images.githubusercontent.com/95432912/145849958-ad4e7caf-6570-40e6-8326-c11e0e0edc50.png) to
![image](https://user-images.githubusercontent.com/95432912/145852163-b3f5a0ee-3244-4de5-840b-32e0c979b98d.png), and ![image](https://user-images.githubusercontent.com/95432912/145850373-a38cc8a8-ff69-4d3c-b1e7-ae1ddf9d7d08.png) to ![image](https://user-images.githubusercontent.com/95432912/145852246-e02160eb-0c81-40d2-83bb-0a7c6c9a14bc.png). This is the part that troubling me because they are not explicitly written in PenaltyInterfaceDiffusion code, so I don't know how to change it.

The equation in master domain will be:
![image](https://user-images.githubusercontent.com/95432912/145851628-a87d6f4e-cdf5-4727-90ff-03f0487a63bd.png)
in the neighbor domain:
![image](https://user-images.githubusercontent.com/95432912/145852053-14fdf86a-df69-490b-87a6-74ac899f3273.png)

I hope this is clear, thank you.
